commit.go

```diff
@@ -1,310 +1,199 @@
+		sha, err := hex.DecodeString(string(line)[16:])
+		if err != nil {
+			return err
 		}
 
-		i, err := strconv.Atoi(path[1])
+		mt, rest, err := readMergeTag(sha, rd)
 		if err != nil {
-			return nil, nil, err
+			return err
 		}
 
-		if i < 0 || i >= len(c.MergeTag) {
-			return nil, nil, fmt.Errorf("index out of range")
-		}
+		c.MergeTag.x = append(c.MergeTag.x, *mt)
 
-		if len(path) == 2 {
-			return c.MergeTag[i], nil, nil
+		if rest != nil {
+			err = decodeCommitLine(c, rest, rd)
+			if err != nil {
+				return err
+			}
+		}
+	case bytes.HasPrefix(line, []byte("gpgsig ")):
+		sig, err := decodeGpgSig(rd)
+		if err != nil {
+			return err
+		}
+		c.Sig = _GpgSig__Maybe{m: schema.Maybe_Value, v: sig}
+	case len(line) == 0:
+		rest, err := ioutil.ReadAll(rd)
+		if err != nil {
+			return err
 		}
-		return c.MergeTag[i].resolve(path[2:])
+
+		c.Message = _String{string(rest)}
 	default:
-		return nil, nil, errors.New("no such link")
+		c.Other.x = append(c.Other.x, _String{string(line)})
 	}
+	return nil
 }
 
-func (c *Commit) ResolveLink(path []string) (*node.Link, []string, error) {
-	out, rest, err := c.Resolve(path)
+func decodeGpgSig(rd *bufio.Reader) (GpgSig, error) {
+	line, _, err := rd.ReadLine()
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	lnk, ok := out.(*node.Link)
-	if !ok {
-		return nil, nil, errors.New("not a link")
+	out := _GpgSig{}
+
+	if string(line) != " " {
+		if strings.HasPrefix(string(line), " Version: ") || strings.HasPrefix(string(line), " Comment: ") {
+			out.x += string(line) + "\n"
+		} else {
+			return nil, fmt.Errorf("expected first line of sig to be a single space or version")
+		}
+	} else {
+		out.x += " \n"
 	}
 
-	return lnk, rest, nil
-}
+	for {
+		line, _, err := rd.ReadLine()
+		if err != nil {
+			return nil, err
+		}
 
-func (c *Commit) Size() (uint64, error) {
-	return uint64(len(c.RawData())), nil
-}
+		if bytes.Equal(line, []byte(" -----END PGP SIGNATURE-----")) {
+			break
+		}
 
-func (c *Commit) Stat() (*node.NodeStat, error) {
-	return &node.NodeStat{}, nil
-}
+		out.x += string(line) + "\n"
+	}
 
-func (c *Commit) String() string {
-	return "[git commit object]"
+	return &out, nil
 }
 
-func (c *Commit) Tree(p string, depth int) []string {
-	if depth != -1 {
-		panic("proper tree not yet implemented")
-	}
-	tree := []string{"tree", "parents", "message", "gpgsig"}
-	tree = append(tree, c.Author.tree("author", depth)...)
-	tree = append(tree, c.Committer.tree("committer", depth)...)
-	for i := range c.Parents {
-		tree = append(tree, fmt.Sprintf("parents/%d", i))
+func encodeCommit(n ipld.Node, w io.Writer) error {
+	c, ok := n.(Commit)
+	if !ok {
+		ci := Type.Commit__Repr.NewBuilder()
+		if err := ci.AssignNode(n); err != nil {
+			return fmt.Errorf("not a Commit: %T %w", n, err)
+		}
+		ciN := ci.Build()
+		c, ok = ciN.(Commit)
 	}
-	return tree
-}
 
-func (c *Commit) GitSha() []byte {
-	return cidToSha(c.Cid())
-}
+	buf := new(bytes.Buffer)
```

Comment:
if you think this buffer could get large, you could also use a bufio.Writer, which will write to the final writer in chunks. But if we think it will never be more than a few kilobytes, it's likely irrelevant.